Update Kubeflow Notebooks docs

thesuperzapper opened the pull request:
__This PR updates the Kubeflow Notebooks docs:__
- creates new "overview" page
- removes specific references to "Jupyter" (as Kubeflow 1.3 now supports multiple IDEs)
- added lots of information about how to build your own container images (for Jupyter, VSCode, and RStudio)
- simplifies the "getting started" guide (and makes it generic to not just be about Jupyter)

__This is the first step towards having the docs updated for 1.4:__
- https://github.com/kubeflow/manifests/issues/1961

__In future PRs, to be updated for 1.3/1.4, we need to:__
- the PNG images of the UI need to be updated for 1.4
- add a page about `spawner_ui_config.yaml`
- add a page about how to configure logos for "group 1" and "group 2"
    - `/src/apps/default/static/assets/group-one-icon.svg`
    - `/src/apps/default/static/assets/group-one-logo.svg`
    - `/src/apps/default/static/assets/group-two-icon.svg`
    - `/src/apps/default/static/assets/jupyter-icon.svg`
    - `/src/apps/default/static/assets/jupyterlab-logo.svg`

thesuperzapper commented:
@RFMVasconcelos or @kubeflow/wg-notebooks-leads can you please review?

thesuperzapper commented:
@zijianjoy can you please review this before we cut the 1.4 branch (it's pretty important updates for the Kubeflow Notebooks docs)!

kimwnasptd commented:
@thesuperzapper yep, it's one of my highest priorities! I'm having my hands full with KubeCon right now, but I'll try to do this in parallel in the next days

shannonbradshaw commented on content/en/docs/components/notebooks/jupyter-tensorflow-examples.md:
```diff
@@ -0,0 +1,59 @@
+
++++
+
+## Mnist Example
```
It would be helpful to have a brief intro paragraph explaining what this paragraph is doing.

shannonbradshaw commented on content/en/docs/components/notebooks/quickstart-guide.md:
```diff
@@ -0,0 +1,113 @@
+
+9. Specify a __"workspace volume"__ to be mounted as a PVC Volume on your home folder.
+
+10. *(Optional)* Specify one or more __"data volumes"__ to be mounted as a PVC Volumes.
```
I'd recommend including a screenshot for workspace and data volumes.

shannonbradshaw commented on content/en/docs/components/notebooks/quickstart-guide.md:
```diff
@@ -0,0 +1,113 @@
+
+10. *(Optional)* Specify one or more __"data volumes"__ to be mounted as a PVC Volumes.
+
+11. *(Optional)* Specify one or more additional __"configurations"__
```
And one for configurations. Given the number of options, it's easy for people to get lost. 

I realize we want to focus users on the required steps, but I think these screenshots add clarity.

shannonbradshaw commented on content/en/docs/components/notebooks/quickstart-guide.md:
```diff
@@ -0,0 +1,113 @@
+    - It can take a few minutes for kubernetes to provision the notebook server pod.
+    - You can check the status of your Pod by hovering your mouse cursor over the icon in the __"Status"__ column.
+
+15. Click __"CONNECT"__ to view the web interface exposed by your notebook server.
```
Nit. Maybe include a red arrow for the Connect link as we did above.

shannonbradshaw commented on content/en/docs/components/notebooks/quickstart-guide.md:
```diff
@@ -0,0 +1,113 @@
+    - Currently, there is no implementation in Kubernetes to activate shared memory.
+    - As a workaround, Kubeflow mounts an empty directory volume at `/dev/shm`.
+
+14. Click __"LAUNCH"__ to create a new Notebook CRD with your specified settings.
```
Since this button is hidden at the bottom until you scroll, I'd suggest a screenshot for this as well.

shannonbradshaw commented:
Left a few comments, but the overall structure and additions look good to me. Would you like to get a review from @kimwnasptd as well.

Thanks @thesuperzapper ! Really appreciate the work here.

kimwnasptd commented on content/en/docs/components/notebooks/jupyter-tensorflow-examples.md:
```diff
@@ -0,0 +1,59 @@
+
+- See a [simple example](https://github.com/kubeflow/examples/tree/master/pipelines/simple-notebook-pipeline) of creating Kubeflow pipelines in a Jupyter notebook.
+- Build machine-learning pipelines with the [Kubeflow Pipelines SDK](/docs/components/pipelines/sdk/sdk-overview/).
+- Explore [Kubeflow Fairing](/docs/external-add-ons/fairing/) for a complete solution to building, training, and deploying an ML model from a notebook.
```
Let's not recommend Fairing here.
It has not been maintained for more than a year and its docs are out of date. It also does not currently belong to any WG

shannonbradshaw replied:
+1

thesuperzapper replied:
Resolved by https://github.com/kubeflow/website/pull/3003/commits/d91928b2d9633cbad49c8a9275a8bc40b5c488c5

kimwnasptd commented on content/en/docs/components/notebooks/jupyter-tensorflow-examples.md:
```diff
@@ -0,0 +1,59 @@
+- See a [simple example](https://github.com/kubeflow/examples/tree/master/pipelines/simple-notebook-pipeline) of creating Kubeflow pipelines in a Jupyter notebook.
+- Build machine-learning pipelines with the [Kubeflow Pipelines SDK](/docs/components/pipelines/sdk/sdk-overview/).
+- Explore [Kubeflow Fairing](/docs/external-add-ons/fairing/) for a complete solution to building, training, and deploying an ML model from a notebook.
+- Learn the advanced features available from a Kubeflow notebook, such as [submitting Kubernetes resources](/docs/components/notebooks/submit-kubernetes/) or [building Docker images](/docs/components/notebooks/custom-notebook/). 
```
Let's rephrase this one a little bit as well:

```
Learn the advanced features available in the example notebook images, such as ...
```

thesuperzapper replied:
As discussed here, https://github.com/kubeflow/website/pull/3003#issuecomment-960241069, we will use "Kubeflow Notebooks"

kimwnasptd reviewed (changes requested):
Thank you very much for this work @thesuperzapper! I've left some initial comments on some parts of your work. The main theme of the comments is to not use the naming `Kubeflow Notebooks` since this implies that only some Notebooks are Kubeflow while others are not.

I'll also do a review of the `overview`, `quickstart-guide`, `submit-kubernetes`, `troubleshooting` pages tomorrow as well.

thesuperzapper commented:
> Thank you very much for this work @thesuperzapper! I've left some initial comments on some parts of your work. The main theme of the comments is to not use the naming `Kubeflow Notebooks` since this implies that only some Notebooks are Kubeflow while others are not.

@kimwnasptd I think I am not understanding what you mean by "this implies that only some Notebooks are Kubeflow while others are not".

In my mind, the name "Kubeflow Notebooks" is the __brand-name__ of the "applications which run notebooks in Kubeflow". Similarly, "Kubeflow Pipelines" is the __brand-name__ of the "set of applications that run pipelines in Kubeflow" (note, we already use "Kubeflow Pipelines" in the docs for that component).

To put it another way, these docs are describing the "set of applications in Kubeflow which run notebooks", and the collective noun for those things is "Kubeflow Notebooks" not "Notebooks".

I feel very strongly that this will improve the brand recognition of the Kubeflow Notebooks "product".

StefanoFioravanzo commented:
@thesuperzapper @kimwnasptd My two cents here on naming. So the problem stems from the fact that the component’s name is the same name as the objects being handled by the component. I think this is why it’s `Kubeflow Pipelines` and not just `Pipelines`. Because `Pipelines` is a generic term that by itself does not identify a specific product. Take `Katib` or `KFServing` instead, they are brand names that identify a specific component.

So, in terms of Docs, I think that using `Kubeflow Notebooks` to refer to the components and `Notebooks` to the specific instances can help. **BUT**, this has in no way change what we do in the dashboard. There, we should continue showing `Notebooks`, which are the **objects**.

thesuperzapper commented:
> @thesuperzapper @kimwnasptd My two cents here on naming. So the problem stems from the fact that the component’s name is the same name as the objects being handled by the component. I think this is why it’s `Kubeflow Pipelines` and not just `Pipelines`. Because `Pipelines` is a generic term that by itself does not identify a specific product. Take `Katib` or `KFServing` instead, they are brand names that identify a specific component.

@StefanoFioravanzo If I am reading this correctly, you are agreeing with my proposal to use the "brand name" of `"Kubeflow Notebooks"` on https://kubeflow.org, rather than what is currently on the website: `"Notebook Servers"`?

> So, in terms of Docs, I think that using `Kubeflow Notebooks` to refer to the components and `Notebooks` to the specific instances can help. **BUT**, this has in no way change what we do in the dashboard. There, we should continue showing `Notebooks`, which are the **objects**.

I completely agree that within Central Dashboard, it makes no sense to specify `"Kubeflow Notebooks"` on the sidebar, and we should continue to use `"Notebooks"`. However, this PR is changing the section-name on https://kubeflow.org from `"Notebook Servers"` to `"Kubeflow Notebooks"` (which will not affect Central Dashboard).

shannonbradshaw commented on content/en/docs/components/notebooks/container-images.md:
```diff
@@ -0,0 +1,73 @@
+
+### Full Images
+
+These images extend the [base images](#base-images) with common packages used in the real world.
```
+1

thesuperzapper replied:
Resolved by https://github.com/kubeflow/website/pull/3003/commits/d91928b2d9633cbad49c8a9275a8bc40b5c488c5

shannonbradshaw commented:
@kimwnasptd you gave a thumbs up to @StefanoFioravanzo comment above. Are you in agreement with @thesuperzapper's summary that follows that comment?

thesuperzapper commented on content/en/docs/components/notebooks/_index.md:
```diff
@@ -1,5 +1,5 @@
 +++
-title = "Notebook Servers"
-description = "Using Jupyter notebooks in Kubeflow"
+title = "Kubeflow Notebooks"
```
As discussed here, https://github.com/kubeflow/website/pull/3003#issuecomment-960241069, we will use "Kubeflow Notebooks"

thesuperzapper commented on content/en/docs/components/notebooks/container-images.md:
```diff
@@ -0,0 +1,73 @@
++++
+title = "Container Images"
+description = "About Container Images for Kubeflow Notebooks"
```
As discussed here, https://github.com/kubeflow/website/pull/3003#issuecomment-960241069, we will use "Kubeflow Notebooks"

thesuperzapper commented on content/en/docs/components/notebooks/_index.md:
```diff
@@ -1,5 +1,5 @@
-title = "Notebook Servers"
-description = "Using Jupyter notebooks in Kubeflow"
+title = "Kubeflow Notebooks"
+description = "Documentation for Kubeflow Notebooks"
```
As discussed here, https://github.com/kubeflow/website/pull/3003#issuecomment-960241069, we will use "Kubeflow Notebooks"

thesuperzapper commented on content/en/docs/components/notebooks/container-images.md:
```diff
@@ -0,0 +1,73 @@
+weight = 30
+                    
++++
+Kubeflow Notebooks natively supports three types of notebooks, [JupyterLab](https://github.com/jupyterlab/jupyterlab), [RStudio](https://github.com/jupyterlab/jupyterlab), and [Visual Studio Code (code-server)](https://github.com/cdr/code-server), but any web-based IDE should work.
```
As discussed here, https://github.com/kubeflow/website/pull/3003#issuecomment-960241069, we will use "Kubeflow Notebooks"

thesuperzapper commented on content/en/docs/components/notebooks/container-images.md:
```diff
@@ -0,0 +1,73 @@
+weight = 30
+                    
++++
+Kubeflow Notebooks natively supports three types of notebooks, [JupyterLab](https://github.com/jupyterlab/jupyterlab), [RStudio](https://github.com/jupyterlab/jupyterlab), and [Visual Studio Code (code-server)](https://github.com/cdr/code-server), but any web-based IDE should work.
```
In this case, we are using the term RStudio to refer to the software itself, which should be fine.

thesuperzapper commented on content/en/docs/components/notebooks/container-images.md:
```diff
@@ -0,0 +1,73 @@
+
+### Base Images
+
+These images provide a common starting point for Kubeflow Notebook containers.
```
As discussed here, https://github.com/kubeflow/website/pull/3003#issuecomment-960241069, we will use "Kubeflow Notebooks"

thesuperzapper commented on content/en/docs/components/notebooks/jupyter-tensorflow-examples.md:
```diff
@@ -0,0 +1,59 @@
++++
+title = "Jupyter TensorFlow Examples"
+description = "Examples using Jupyter and TensorFlow in Kubeflow Notebooks"
```
As discussed here, https://github.com/kubeflow/website/pull/3003#issuecomment-960241069, we will use "Kubeflow Notebooks"

thesuperzapper commented on content/en/docs/components/notebooks/container-images.md:
```diff
@@ -0,0 +1,73 @@
+
+Packages that are installed by users __after spawning__ a notebook server will __only last the lifetime of the Pod__ unless those packages are installed into a PVC-backed directory.
+
+We recommend extending the [base images](#base-images) with the packages and dependencies you need. (See [a guide](https://github.com/kubeflow/kubeflow/tree/master/components/example-notebook-servers#custom-images) that explains how to do this.)
```
Resolved by https://github.com/kubeflow/website/pull/3003/commits/d91928b2d9633cbad49c8a9275a8bc40b5c488c5

thesuperzapper commented on content/en/docs/components/notebooks/container-images.md:
```diff
@@ -0,0 +1,73 @@
+weight = 30
+                    
++++
+Kubeflow Notebooks natively supports three types of notebooks, [JupyterLab](https://github.com/jupyterlab/jupyterlab), [RStudio](https://github.com/jupyterlab/jupyterlab), and [Visual Studio Code (code-server)](https://github.com/cdr/code-server), but any web-based IDE should work.
```
Resolved by https://github.com/kubeflow/website/pull/3003/commits/d91928b2d9633cbad49c8a9275a8bc40b5c488c5

thesuperzapper commented:
@shannonbradshaw thanks for your comments, but in the interest of getting this PR merged, can you raise a separate PR which makes your desired changes after this has been merged (NOTE: the `notebooks/quickstart-guide.md` is not actually a new page, it was just moved/cleaned by this PR)

thesuperzapper commented:
@Bobgy this PR has been reviewed by @shannonbradshaw and @kimwnasptd and I have addressed the comments from them, but we still need someone with root OWNER to `/approve` this (because it updates the `_redirects` file)

Can you `/approve` this so that we can have the notebooks docs updated for Kubeflow 1.4 (and 1.3, lol)?

shannonbradshaw commented:
@thesuperzapper it is OK with me to address my comments in a separate PR. Thanks for following up. It looks like we've provided enough time for other reviewers to chime back in. I'm in favor of merging this. @Bobgy or @zijianjoy

Bobgy commented:
/approve
I'm interested to see if there are ways for WGs to self service on redirection changes.

Looks like hugo also supports a field called "aliases", it can be added directly on content pages.
Can you confirm if you can set up redirect in that approach?
https://gohugo.io/content-management/urls/#example-aliases

thesuperzapper commented:
@shannonbradshaw can you `/lgtm` so this gets merged?

kimwnasptd commented:
I'll go on and merge this PR, thanks for your effort in this @thesuperzapper!

Although, in the future please give an explicit heads up to WG leads when renaming the product name of a Component since this needs discussion. This is a significant change and there was no mention of this the PR description. Doing such a change alongside a whole refactoring makes it difficult to discuss specific parts when dealing with such huge PRs.

/lgtm
